### ISSUE TITLE
Support identifying the crate that's holding a dependency at a vulnerable version

### ISSUE THREAD

**Issue by ssokolow** [RustSec/cargo-audit#306]
Currently, when a deep transitive dependency has a vulnerability and the solution is to upgrade, cargo-audit doesn't provide any help identifying what point in the dependency chain is preventing that from happening.

cargo-audit should at least have the option to walk up the dependency tree from the vulnerable package to the packages listed in `Cargo.toml` and identify the crates where, if they pushed a new non-semver-breaking update, they could resolve the problem, so users easily know which projects to file bug reports on and/or contribute PRs to.

As an example, here's one I ran into:

<details>

```
Crate:         smallvec
Version:       1.6.0
Title:         Buffer overflow in SmallVec::insert_many
Date:          2021-01-08
ID:            RUSTSEC-2021-0003
URL:           https://rustsec.org/advisories/RUSTSEC-2021-0003
Solution:      Upgrade to >=0.6.14, <1.0.0 OR >=1.6.1
Dependency tree: 
smallvec 1.6.0
├── trust-dns-resolver 0.19.6
│   └── actix-connect 2.0.0
│       └── actix-http 2.2.0
│           ├── awc 2.0.3
│           │   └── actix-web 3.3.2
│           │       ├── [my project] 0.1.0
│           │       ├── actix-web-httpauth 0.5.0
│           │       │   └── [my project] 0.1.0
│           │       └── actix-files 0.5.0
│           │           └── [my project] 0.1.0
│           └── actix-web 3.3.2
├── trust-dns-proto 0.19.6
│   ├── trust-dns-resolver 0.19.6
│   └── actix-connect 2.0.0
├── parking_lot_core 0.8.2
│   └── parking_lot 0.11.1
│       └── actix-threadpool 0.3.3
│           ├── actix-web 3.3.2
│           ├── actix-rt 1.1.1
│           │   ├── [my project] 0.1.0
│           │   ├── awc 2.0.3
│           │   ├── actix-web 3.3.2
│           │   ├── actix-utils 2.0.0
│           │   │   ├── actix-web 3.3.2
│           │   │   ├── actix-tls 2.0.0
│           │   │   │   └── actix-web 3.3.2
│           │   │   ├── actix-server 1.0.4
│           │   │   │   ├── actix-web 3.3.2
│           │   │   │   └── actix-testing 1.0.1
│           │   │   │       └── actix-web 3.3.2
│           │   │   ├── actix-http 2.2.0
│           │   │   └── actix-connect 2.0.0
│           │   ├── actix-testing 1.0.1
│           │   ├── actix-server 1.0.4
│           │   ├── actix-http 2.2.0
│           │   └── actix-connect 2.0.0
│           └── actix-http 2.2.0
└── actix-rt 1.1.1
```

</details>

How is it supposed to be practical for a hobby developer to figure out a resolution for this other than running `cargo upgrade --allow-prerelease` to upgrade to a beta version of actix-web and hoping for the best?

**Comment by tarcieri** [comment]
> How is it supposed to be practical for a hobby developer to figure out a resolution for this

This is the sort of thing `cargo audit fix` should probably do, although right now its logic is too naive to address this particular situation.

Given this was the resolution:

> ...other than running cargo upgrade --allow-prerelease to upgrade to a beta version of actix-web and hoping for the best?

...it seems like a tricky situation in general, and perhaps one that's best addressed by the upstream project opening a tracking ticket with instructions for its users on how to address the issue.

**Comment by ssokolow** [comment]
> ...it seems like a tricky situation in general, and perhaps one that's best addressed by the upstream project opening a tracking ticket with instructions for its users on how to address the issue.

My point is that it seems like a very crude instrument and possibly unfair to expect the top of a dependency hierarchy to always be the point of contact for this sort of thing when they might already be depending on the newest version of something in between them and the vulnerable dependency.

Hence my suggestion of something to automate the process of finding closer points in the dependency chain that could solve the problem.

(eg. in the example, what if it turns out that all of Actix's dependency requests are already allowing smallvec v1.6.1 but it's a trust-dns-* crate that's saying "No, it has to be 1.6.0. I won't accept 1.6.1."?)

**Comment by tarcieri** [comment]
I agree, and this is what `cargo audit fix` is for, but this is a much more complicated situation than it is currently equipped to handle

**Comment by ssokolow** [comment]
Fair enough. Also, I just realized that I didn't address "Given this was the resolution:"

I should point out that, for this (not yet released) project, I haven't actually tried upgrading to Actix beta... it's just that I can see a lot of people (myself included, when I'm too tired to be properly motivated) seeing "allow beta releases of crates I depend on, consider that my due diligence, and if that doesn't fix it, rationalize that the vulnerability must not be in code that Actix actually calls" as the solution for hobby projects when the alternative is to dig down into the transitive dependencies... especially when we've never done any local dependency patching before.
